### PR TITLE
[shopsys] multiple delivery addresses

### PR DESCRIPTION
**Pull request by TomasLudvik**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Some customers prefer to store multiple delivery addresses, this PR enables such functionality.
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| Yes 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/src/Form/Constraints/DeliveryAddressOfCurrentCustomerValidator.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+     * @param \Symfony\Component\Validator\Constraint $constraint
+     */
+    public function validate($deliveryAddress, Constraint $constraint): void
+    {
+        if (!$constraint instanceof DeliveryAddressOfCurrentCustomer) {
+            throw new \Symfony\Component\Validator\Exception\UnexpectedTypeException($constraint, UniqueCollection::class);
+        }
+
+        if ($deliveryAddress !== null) {
+            if ($this->currentCustomerUser->findCurrentCustomerUser() !== null) {
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Form/Constraints/DeliveryAddressOfCurrentCustomerValidator.php]: Done

File: packages/framework/src/Model/Customer/Customer.php [hunk 2]
```diff
@@ -29,15 +29,35 @@ class Customer
...
+        $this->deliveryAddresses = new ArrayCollection();
+    }
+
     /**
      * @param \Shopsys\FrameworkBundle\Model\Customer\BillingAddress $billingAddress
      */
     public function addBillingAddress(BillingAddress $billingAddress): void
     {
-        $this->billingAddresses = new ArrayCollection();
         $this->billingAddresses->add($billingAddress);
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Model/Customer/Customer.php]: I agree

File: packages/framework/src/Model/Customer/DeliveryAddressData.php [hunk 3]
```diff
@@ -49,6 +49,11 @@ class DeliveryAddressData
      */
     public $country;
 
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Customer\Customer
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Model/Customer/DeliveryAddressData.php]: fixed

File: packages/framework/src/Model/Customer/DeliveryAddressFacade.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        $this->em = $em;
+    }
+
+    /**
+     * @param int $deliveryAddressId
+     * @param \Shopsys\FrameworkBundle\Model\Customer\DeliveryAddressData $deliveryAddressData
+     */
+    public function edit(int $deliveryAddressId, DeliveryAddressData $deliveryAddressData): void
+    {
+        $deliveryAddress = $this->deliveryAddressRepository->getById($deliveryAddressId);
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Model/Customer/DeliveryAddressFacade.php]: Done

File: packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php [hunk 5]
```diff
@@ -72,22 +73,22 @@ public function createFromCustomerUser(CustomerUser $customerUser): CustomerUser
...
+     * @param \Shopsys\FrameworkBundle\Model\Customer\Customer $customer
      *
      * @return \Shopsys\FrameworkBundle\Model\Customer\DeliveryAddressData
      */
-    protected function getDeliveryAddressDataFromCustomerUser(CustomerUser $customerUser): DeliveryAddressData
+    protected function getDeliveryAddressDataFromCustomer(Customer $customer): DeliveryAddressData
     {
-        if ($customerUser->getDeliveryAddress()) {
-            return $this->deliveryAddressDataFactory->createFromDeliveryAddress($customerUser->getDeliveryAddress());
+        if ($customer->getLastCreatedDeliveryAddress()) {
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Model/Customer/User/CustomerUserUpdateDataFactory.php]: fixed

File: packages/framework/tests/Unit/Model/Customer/CustomerUserUpdateDataFactoryTest.php [hunk 6]
```diff
@@ -76,10 +76,14 @@ public function testGetAmendedCustomerUserUpdateDataByOrderWithoutChanges()
...
         $deliveryAddressData->telephone = 'deliveryTelephone';
         $deliveryAddressData->country = $deliveryCountry;
+        $deliveryAddressData->customer = $customer;
 
         $customer->addBillingAddress($this->createBillingAddress($billingAddressData));
-        $deliveryAddress = $this->createDeliveryAddress($deliveryAddressData);
-        $customerUser = new CustomerUser($customerUserData, $deliveryAddress);
+        $customer->addDeliveryAddress($this->createDeliveryAddress($deliveryAddressData));
+
+        $deliveryAddress = new DeliveryAddress($deliveryAddressData);
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/tests/Unit/Model/Customer/CustomerUserUpdateDataFactoryTest.php]: fixed

File: project-base/assets/js/frontend/deliveryAddress/deliveryAddress.js [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                    deliveryAddress.onDeliveryAddressChange($deliveryAddressSelect, deliveryAddress, $this);
+                },
+                error: function () {
+                    deliveryAddress.errorMessage();
+                }
+            });
+        }
+    }
+
+    successMessage () {
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/assets/js/frontend/deliveryAddress/deliveryAddress.js]: renamed

File: packages/framework/src/Resources/views/Front/Form/deliveryAddressChoiceFields.html.twig [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                                    <a data-href="{{ url('front_customer_delivery_address_delete') }}" class="js-delivery-address-remove-button">
+                                        <i class="svg svg-remove"></i>
+                                    </a>
+                                </td>
+                            </tr>
+                        {% else %}
+                            <tfoot>
+                                <tr>
+                                    <td>{{ form_widget(child, { attr: { class: 'js-delivery-address-input' } }) }}</td>
+                                    <td colspan="9"><label for="{{ child.vars.id }}">{{ 'New shipping address'|trans }}</label></td>
```
Comment (DavidOstrozlik) [inline packages/framework/src/Resources/views/Front/Form/deliveryAddressChoiceFields.html.twig]: you are sometimies calling it shipping and sometimes delivery address so please change it everywhere to delivery address

File: project-base/assets/js/frontend/deliveryAddress/deliveryAddress.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+        // eslint-disable-next-line no-new
+        new Window({
+            content: Translator.trans('Do you really want to remove this shipping address?'),
+            buttonCancel: true,
+            buttonContinue: true,
+            eventContinue: () => {
+                const deliveryAddressId = $input.val();
+                if (deliveryAddressId > 0) {
+                    const url = $this.data('href') + '/' + deliveryAddressId;
```
Comment (DavidOstrozlik) [inline project-base/assets/js/frontend/deliveryAddress/deliveryAddress.js]: why dont you just create the right url with the right parameter in the template?

File: project-base/config/routes/shopsys_front.yml [hunk 10]
```diff
@@ -186,3 +186,13 @@ front_download_uploaded_file:
     methods: [GET]
     requirements:
         uploadedFileId: \d+
+
+front_customer_delivery_address_delete:
+    path: /customer/delete-delivery-address/{deliveryAddressId}
+    defaults:
+        _controller: App\Controller\Front\CustomerController:deleteDeliveryAddressAction
+        deliveryAddressId: 0
```
Comment (DavidOstrozlik) [inline project-base/config/routes/shopsys_front.yml]: why are you using default deliveryAddressId?

File: project-base/templates/Front/Content/Customer/edit.html.twig [hunk 11]
```diff
@@ -75,21 +75,24 @@
...
                         <div id="js-delivery-address-fields">
-                            {{ form_row(form.deliveryAddressData.companyName, { label: 'Company'|trans}) }}
-                            {{ form_row(form.deliveryAddressData.firstName, { label: 'First name'|trans}) }}
-                            {{ form_row(form.deliveryAddressData.lastName, { label: 'Last name'|trans}) }}
-                            {{ form_row(form.deliveryAddressData.telephone, { label: 'Telephone'|trans}) }}
-                            {{ form_row(form.deliveryAddressData.street, { label: 'Street'|trans}) }}
-                            {{ form_row(form.deliveryAddressData.city, { label: 'City'|trans}) }}
-                            {{ form_row(form.deliveryAddressData.postcode, { label: 'Postcode'|trans}) }}
-                            {{ form_row(form.deliveryAddressData.country, { label: 'Country'|trans}) }}
+                            {{ form_row(form.deliveryAddressData.companyName, { label: 'Company'|trans, attr: { class: 'js-delivery-address-companyName' } }) }}
```
Comment (DavidOstrozlik) [inline project-base/templates/Front/Content/Customer/edit.html.twig]: these js-* classes doesn't seem to be used. Do they need to be here?

File: packages/framework/src/Resources/views/Front/Form/deliveryAddressChoiceFields.html.twig (hunk 8 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/views/Front/Form/deliveryAddressChoiceFields.html.twig]: Well displayed name in administration and frontend is Shipping address, but in code it is always called delivery address... I know... confusing, but that is how it is

File: project-base/templates/Front/Content/Customer/edit.html.twig (hunk 11 above)
Comment (TomasLudvik) [reply project-base/templates/Front/Content/Customer/edit.html.twig]: fixed

File: project-base/assets/js/frontend/deliveryAddress/deliveryAddress.js (hunk 9 above)
Comment (TomasLudvik) [reply project-base/assets/js/frontend/deliveryAddress/deliveryAddress.js]: fixed

File: project-base/config/routes/shopsys_front.yml (hunk 10 above)
Comment (TomasLudvik) [reply project-base/config/routes/shopsys_front.yml]: fixed

File: packages/framework/src/Resources/views/Front/Form/deliveryAddressChoiceFields.html.twig (hunk 8 above)
Comment (DavidOstrozlik) [reply packages/framework/src/Resources/views/Front/Form/deliveryAddressChoiceFields.html.twig]: so can you please change it to shipping address in all places? you have it as delivery adderess in eg. deliveryAddressListFields.html.twig

Comment (TomasLudvik) [reply packages/framework/src/Resources/views/Front/Form/deliveryAddressChoiceFields.html.twig]: I have united the naming

Comment (TomasLudvik) [reply packages/framework/src/Resources/views/Front/Form/deliveryAddressChoiceFields.html.twig]: Also I found, that better is to use delivery address as described in commit message:

https://www.proz.com/kudoz/english/business-commerce-general/2766359-shipping-address-and-delivery-address.html
Explanation:
shipping address implies the port where the consignment is unloaded (e.g. FOB Hamburg) whereas delivery address refers to the place the consignee or addressee is resident in. The goods are unloaded in Hamburg from where they will be dispatched to the delivery address.

File: packages/framework/src/Resources/translations/validators.cs.po [hunk 12]
```diff
@@ -379,6 +379,9 @@ msgstr "Stejný doplněk titulku už je používán na jiné doméně"
 msgid "Same title is used on another domain"
 msgstr "Stejný titulek už je používán na jiné doméně"
 
+msgid "Selected delivery address no longer exists"
+msgstr ""
```
Comment (DavidOstrozlik) [inline packages/framework/src/Resources/translations/validators.cs.po]:
```suggestion
msgstr "Vybraná doručovací adresa neexistuje"
```

File: project-base/translations/messages.cs.po [hunk 13]
```diff
@@ -31,6 +31,9 @@ msgstr "Uplatnit"
 msgid "Attention! You are administrator logged in as the customer."
 msgstr "Pozor! Jste jako administrátor přihlášen za zákazníka."
 
+msgid "Availability"
+msgstr ""
```
Comment (DavidOstrozlik) [inline project-base/translations/messages.cs.po]: could you please translate these messages?

File: packages/framework/src/Resources/translations/validators.cs.po (hunk 12 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/translations/validators.cs.po]: fixed

File: project-base/translations/messages.cs.po (hunk 13 above)
Comment (TomasLudvik) [reply project-base/translations/messages.cs.po]: fixed

**Comment by romankuna** [comment]
Hi @TomasLudvik , please check and fix this problems
1. when i made order as logged user, and go tu user section, there is server error id 2VBdPD8Chl
2. whe i made order as logged user, and o go to cart on third step, there is o server error id f8hTLdZydh

**Comment by TomasLudvik** [comment]
Hi @romankuna,

I have fixed everything you have mentioned.

**Comment by romankuna** [comment]
> Hi @romankuna,
> 
> I have fixed everything you have mentioned.

Hi @TomasLudvik  ,

1. ok
2. ok
3. I prefer to put a break between delivery adress and checkbox - user section http://prntscr.com/r2bsp3 
4. when i have choosen delivery adress, there is a button for delete, but i can't delete it. http://prntscr.com/r2bw28 - I prefere not to show thedelete button 
5. when i saved a new delivery adress without required values in user section, the error message is not correct - i thing it should be in popup http://prntscr.com/r2bzfa 
6. i prefer to put break up firtst name input in oder http://prntscr.com/r2c2en 
7. list of delivery adress isn't responsive http://prntscr.com/r2c8w6

**Comment by TomasLudvik** [comment]
Hello @romankuna,

I have fixed deleting of addresses, rest will be solved by coder.

**Comment by romankuna** [comment]
Hi @TomasLudvik ,

4) ok

So I waiting for coder

**Comment by romankuna** [comment]
Hi @TomasLudvik 

coder fix all problems, now i thing it's ok.

**Pull request merged**